src/Google.Datastore.V1Beta3/ArrayValuePartial.cs

```diff
@@ -0,0 +1,82 @@
+        public static implicit operator ArrayValue(DateTimeOffset[] values) => ToArrayValue(values, x => x);
+        public static implicit operator ArrayValue(DateTimeOffset?[] values) => ToArrayValue(values, x => x);
+
+        private static ArrayValue ToArrayValue<T>(T[] values, Func<T, Value> converter) =>
+            values == null ? null : new ArrayValue { Values = { values.Select(x => converter(x) ?? Value.ForNull()) } };
+
+        public static explicit operator string[](ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (string) x);
+        public static explicit operator long[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (long)x);
+        public static explicit operator long?[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (long?)x);
+        public static explicit operator Key[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (Key)x);
+        public static explicit operator double[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (double)x);
+        public static explicit operator double?[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (double?)x);
+        public static explicit operator bool[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (bool)x);
+        public static explicit operator bool?[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (bool?)x);
+        public static explicit operator ArrayValue[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (ArrayValue)x);
+        public static explicit operator ByteString[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (ByteString)x);
+        public static explicit operator byte[][] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (byte[])x);
+        public static explicit operator Entity[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (Entity)x);
+        public static explicit operator LatLng[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (LatLng)x);
+        public static explicit operator Timestamp[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (Timestamp)x);
+        public static explicit operator DateTime[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (DateTime)x);
+        public static explicit operator DateTime?[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (DateTime?)x);
+        public static explicit operator DateTimeOffset[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (DateTimeOffset)x);
+        public static explicit operator DateTimeOffset?[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (DateTimeOffset?)x);
+
+        private static T[] FromArrayValue<T>(ArrayValue arrayValue, Func<Value, T> converter) =>
+            arrayValue == null ? null : arrayValue.Values.Select(x => converter(x)).ToArray();
+
+        public IEnumerator<Value> GetEnumerator() => Values.Select(x => x.OrNull()).GetEnumerator();
```

Comment:
Yes, I definitely like the ability to call Add(null), and I see now that this preserves the symmetry of that, so I am fine with it.

I had another thought (although perhaps this is a bit too hacky): maybe have == and != operators for Value which equate null and the NullValue. With this, you could have idiomatic null checks which actually work. However, to make something like this complete, you would also need the Equals override to treat them as equal, and since it is auto-generated, this might not be possible.